packages/alfa-rules/src/helpers/element-checker.ts

```diff
@@ -0,0 +1,142 @@
+import { getRole, Role } from "@siteimprove/alfa-aria";
+import { BrowserSpecific } from "@siteimprove/alfa-compatibility";
+import { Device } from "@siteimprove/alfa-device";
+import {
+  // Element,
+  getElementNamespace,
+  getInputType,
+  InputType,
+  isElement,
+  Namespace,
+  Node
+} from "@siteimprove/alfa-dom";
+import { Option } from "@siteimprove/alfa-util";
+// import { Predicate } from "@siteimprove/alfa-util";
+
+function emptyOrHas<T>(elt: T, set?: Set<T>) {
+  return set === undefined || set.has(elt);
+}
+
+export class ElementChecker {
+  constructor(namesSet?: Set<string>, inputTypeSet?: Set<InputType>) {
+    this.namesSet = namesSet;
+    this.inputTypeSet = inputTypeSet;
+  }
+
+  public withContext(context: Node): ElementCheckerWithContext {
+    return new ElementCheckerWithContext(
+      context,
+      this.namesSet,
+      this.inputTypeSet
+    );
+  }
+
+  protected namesSet?: Set<string>;
+  public withName(...names: Array<string>): ElementChecker {
+    this.namesSet = new Set(names);
+    return this;
+  }
+
+  protected inputTypeSet?: Set<InputType>;
+  public withInputType(...inputTypes: Array<InputType>): ElementChecker {
+    this.inputTypeSet = new Set(inputTypes);
+    return this;
+  }
+
+  public build(): (node: Node) => boolean | BrowserSpecific<boolean> {
+    return node => this.evaluate(node);
+  }
+
+  public evaluate(node: Node): boolean | BrowserSpecific<boolean> {

```

Comment:
1. The topmost `evaluate` always return `boolean` but the inherited one can return a Browser Specific one. I'd like to have a way to say "this can be both because of inheritance, but if calling from this class, it's always `boolean`" but it seems not really possible. This ripples into a bunch of cast (`as boolean`) when using the toplevel method…
2. I'd also love to be able to refine the type (from `node` to `element`) here, but that does not seem possible with browser specific booleans…